lib/Db/TagMapper.php

```diff
@@ -0,0 +1,234 @@
+			);
+		$result = $this->findEntity($qb);
+		return $result;
+	}
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTagForUser(int $id, string $userId): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+					$qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)),
+					$qb->expr()->eq('user_id', $qb->createNamedParameter($userId))
+				);
+		$result = $this->findEntity($qb);
+		return $result;
+	}
+
+
+	/**
+	 * @param integer $id
+	 * @param string $userId
+	 * @return Entity
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function getTag(int $id): Entity {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where($qb->expr()->eq('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT)));
+		$result = $this->findEntity($qb);
+		return $result;
+	}
+
+	/**
+	 * Tag a message in the DB
+	 *
+	 * To tag (flag) a message on IMAP, @see OCA\Mail\Service\MailManager::tagMessage
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @param string $userId
+	 * @return void
+	 */
+	public function tagMessage(Tag $tag, string $messageId, string $userId) {
+		/** @var Tag $exists */
+		try {
+			$exists = $this->getTagByImapLabel($tag->getImapLabel(), $userId);
+			$tag->setId($exists->getId());
+		} catch (DoesNotExistException $e) {
+			$tag = $this->insert($tag);
+		}
+
+		$qb = $this->db->getQueryBuilder();
+		$qb->insert('mail_message_tags');
+		$qb->setValue('imap_message_id', $qb->createNamedParameter($messageId));
+		$qb->setValue('tag_id', $qb->createNamedParameter($tag->getId(), IQueryBuilder::PARAM_INT));
+		$qb->execute();
+	}
+
+	/**
+	 * Remove a tag from a DB message
+	 *
+	 * @param Tag $tag
+	 * @param string $messageId
+	 * @return void
+	 */
+	public function untagMessage(Tag $tag, string $messageId) {
+		$qb = $this->db->getQueryBuilder();
+		$qb->delete('mail_message_tags')
+			->where($qb->expr()->eq('imap_message_id', $qb->createNamedParameter($messageId)))
+			->where($qb->expr()->eq('tag_id', $qb->createNamedParameter($tag->getId())));
+		$qb->execute();
+	}
+
+	/**
+	 * @param array $messages
+	 * @return array
+	 */
+	public function getAllTagsForMessages(array $messages):array {
+		$ids = array_map(function (Message $message) {
+			return $message->getMessageId();
+		}, $messages);
+
+		$qb = $this->db->getQueryBuilder();
+		$ids = $qb->select('mt.*')
+			->from('mail_message_tags', 'mt')
+			->where(
+				$qb->expr()->in('imap_message_id', $qb->createNamedParameter($ids, IQueryBuilder::PARAM_STR_ARRAY))
+			);
+		$qb = $qb->execute();
+		$queryResult = $qb->fetchAll();
+		if (empty($queryResult)) {
+			return [];
+		}
+
+		foreach ($queryResult as $qr) {
+			$result[$qr['imap_message_id']][] = $this->getTag((int)$qr['tag_id']);
```

Comment:
`$result` is uninitalized and so is the index `$qr['imap_message_id']` when you first access it